File: Leetcode 1 - two-sum/1.two-sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 */

// @lc code=start
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  vector<int> twoSum(vector<int> &nums, int target) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    unordered_map<int, int> count;

    for (int i = 0; i < nums.size(); i++) {
      int check = target - nums[i];
      if (count.find(check) != count.end()) {
        return {count[check], i};
      }
      count[nums[i]] = i;
    }

    return {-1, -1};
  }
};

int main() {
  Solution sol;

  vector<int> nums1 = {2, 7, 11, 15};
  vector<int> result1 = sol.twoSum(nums1, 9); // {0, 1}
  cout << "{";
  for (int i = 0; i < result1.size(); i++) {
    cout << result1[i];
    if (i != result1.size() - 1) {
      cout << ", ";
    }
  }
  cout << "}";
  cout << endl;

  vector<int> nums2 = {3, 2, 4};
  vector<int> result2 = sol.twoSum(nums2, 6); // {1, 2}
  cout << "{";
  for (int i = 0; i < result2.size(); i++) {
    cout << result2[i];
    if (i != result2.size() - 1) {
      cout << ", ";
    }
  }
  cout << "}";
  cout << endl;

  vector<int> nums3 = {3, 3};
  vector<int> result3 = sol.twoSum(nums3, 6); // {0, 2}
  cout << "{";
  for (int i = 0; i < result3.size(); i++) {
    cout << result3[i];
    if (i != result3.size() - 1) {
      cout << ", ";
    }
  }
  cout << "}";
}
// @lc code=end

File: Leetcode 1 - two-sum/1.two-sum.java
/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start

import java.util.HashMap;

class Solution {

  public int[] twoSum(int[] nums, int target) {
    HashMap<Integer, Integer> map = new HashMap<>();

    // Check all the numbers in the array
    for (int i = 0; i < nums.length; i++) {
      int remain = target - nums[i];
      // Check if the remain is in the map
      if (map.containsKey(remain)) {
        return new int[] { map.get(remain), i };
      }
      // If the remain is not in the map, put the number and its index in the map
      map.put(nums[i], i);
    }

    return new int[] {};
  }

  public static void main(String[] args) {
    Solution sol = new Solution();

    int[] result1 = sol.twoSum(new int[] { 2, 7, 11, 15 }, 9);
    System.out.println("[" + result1[0] + ", " + result1[1] + "]"); // [0, 1]
    int[] result2 = sol.twoSum(new int[] { 3, 2, 4 }, 6);
    System.out.println("[" + result2[0] + ", " + result2[1] + "]"); // [1, 2]
    int[] result3 = sol.twoSum(new int[] { 3, 3 }, 6);
    System.out.println("[" + result3[0] + ", " + result3[1] + "]"); // [0, 1]
  }
}
// @lc code=end
